### PR TITLE
Add support for Kafka as a notifications target (#2869)

### LINKED ISSUE #2869: [notifications] Add kafka support for bucket notifications.

### PR DESCRIPTION
**Pull request by donatello**
## Description

Same as title

## Motivation and Context

Fixes #2869

## How Has This Been Tested?

Testing was done with a Kafka setup and a Kafka consumer to check that events are sent to Kafka. Instructions to reproduce the tests are here: https://gist.github.com/donatello/a6b5ac908b5b7a4e6fb233e84608bb28

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/config-migrate.go [hunk 1]
```diff
@@ -633,3 +637,91 @@ func migrateV9ToV10() error {
 	)
 	return nil
 }
+
+// Version '10' to '11' migration. Add support for Kafka
+// notifications.
+func migrateV10ToV11() error {
```
Comment (harshavardhana) [inline cmd/config-migrate.go:643]: You need to add a test for this.

File: cmd/notify-kafka.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		)
+	}
+
+	return kafkaConn{p, kn.Topic}, nil
+}
+
+func newKafkaNotify(accountID string) (*logrus.Logger, error) {
+	kafkaNotifyCfg := serverConfig.GetKafkaNotifyByID(accountID)
+
+	// Dial kafka
```
Comment (harshavardhana) [inline cmd/notify-kafka.go]: This comment just names the function name should be more descriptive.

File: cmd/notify-kafka.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+type kafkaNotify struct {
+	Enable bool `json:"enable"`
+
+	Brokers []string `json:"brokers"`
+
+	Topic string `json:"topic"`
+}
+
+type kafkaConn struct {
```
Comment (harshavardhana) [inline cmd/notify-kafka.go:44]: Comments for each struct.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/config-migrate.go (hunk 1 above)
Comment (donatello) [reply cmd/config-migrate.go:643]: Updated the test in `config-migrate_test.go` to include this migration.

File: cmd/notify-kafka.go (hunk 2 above)
Comment (donatello) [reply cmd/notify-kafka.go]: Done

File: cmd/notify-kafka.go (hunk 3 above)
Comment (donatello) [reply cmd/notify-kafka.go:44]: Done.

File: cmd/config-migrate.go [hunk 4]
```diff
@@ -633,3 +637,91 @@ func migrateV9ToV10() error {
...
+	cv10, err := loadConfigV10()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("Unable to load config version ‘9’. %v", err)
```
Comment (vadmeste) [inline cmd/config-migrate.go]: You mean version 10 and not 9 here.

Comment (donatello) [reply cmd/config-migrate.go]: Good catch. This is fixed now.

**Review by harshavardhana — APPROVED** [review]

**Review by krisis — APPROVED** [review]
LGTM

**Review by vadmeste — APPROVED** [review]

**Pull request merged**